src/parser/jobs/blm/modules/RotationWatchdog/ExpectedFireSpellsEvaluator.tsx

```diff
@@ -0,0 +1,104 @@
+import React from 'react'
+import {ROTATION_ERRORS, DEFAULT_SEVERITY_TIERS, CycleMetadata} from '../RotationWatchdog'
+import {assignErrorCode, getMetadataForWindow, includeInSuggestions} from './EvaluatorUtilities'
+
+export type ExpectedFireSpellsEvaluatorOpts =
+	& Omit<TrackedActionsOptions, 'suggestionIcon' | 'suggestionContent' | 'suggestionWindowName' | 'severityTiers'>
+	& {
+	pullEnd: number
+	despairAction: Action
+	fire4Action: Action
+	invulnerability: Invulnerability
+	metadataHistory: History<CycleMetadata>
+}
+
+export class ExpectedFireSpellsEvaluator extends ExpectedActionsEvaluator {
+	private pullEnd: number
+	private despairAction: Action
+	private fire4Action: Action
+	private invulnerability: Invulnerability
+	private metadataHistory: History<CycleMetadata>
+
+	constructor(opts: ExpectedFireSpellsEvaluatorOpts) {
+		super({...opts,
+			suggestionIcon: '',
+			suggestionContent: <></>,
+			suggestionWindowName: <></>,
+			severityTiers: {},
+		})
+		this.pullEnd = opts.pullEnd
+		this.despairAction = opts.despairAction
+		this.fire4Action = opts.fire4Action
+		this.invulnerability = opts.invulnerability
+		this.metadataHistory = opts.metadataHistory
+	}
+
+	// We only actually have a suggestion for missing Despair, but we'll assign the missing Fire 4 codes here too
+	public override suggest(windows: Array<HistoryEntry<EvaluatedAction[]>>) {
+		const astralFiresMissingDespairs = windows.reduce((total, window) => {
+			const windowMetadata = getMetadataForWindow(window, this.metadataHistory)
+
+			this.expectedActions.forEach(action => this.assessWindowAction(window, windowMetadata, action))
+
+			return total + ((windowMetadata.missingDespairs && includeInSuggestions(windowMetadata)) ? 1 : 0)
+		}, 0)
+
+		return new TieredSuggestion({
+			icon: this.despairAction.icon,
+			content: <Trans id="blm.rotation-watchdog.suggestions.end-with-despair.content">
+				Once you can no longer cast another spell in Astral Fire and remain above 800 MP, you should use your remaining MP by casting <DataLink action="DESPAIR"/>.
+			</Trans>,
+			tiers: DEFAULT_SEVERITY_TIERS,
+			value: astralFiresMissingDespairs,
+			why: <Trans id="blm.rotation-watchdog.suggestions.end-with-despair.why">
+				<Plural value={astralFiresMissingDespairs} one="# Astral Fire phase was" other="# Astral Fire phases were"/> missing at least one <DataLink showIcon={false} action="DESPAIR"/>.
+			</Trans>,
+		})
+	}
+
+	override determineExpected(window: HistoryEntry<EvaluatedAction[]>, action: TrackedAction) {
+		const windowMetadata = getMetadataForWindow(window, this.metadataHistory)
+
+		if (action.action.id === this.fire4Action.id) {
+			if (windowMetadata.finalOrDowntime) { return undefined }
+			if (windowMetadata.expectedFire4s >= 0) { return windowMetadata.expectedFire4s }
+		}
+		if (action.action.id === this.despairAction.id && windowMetadata.expectedDespairs >= 0) {
+			return windowMetadata.expectedDespairs
+		}
```

Comment:
I'm not super on board with exposing determineExpected for this logic - it feels like this should be part of the adjustExpectedActionsCount function instead, since really all this is doing is using the logic from that function to return early and never call through to super.determineExpected.  Doubly so since we're defining the base expected per window as 0 for both actions anyway - the adjustment value being returned by adjustExpectedActionsCount is going to wind up being the final value returned anyway except for the case of marking F4 undefined in the last window before downtime.  I think we'd be better off to look at the interface of the base class and change it as necessary so that any consumer can mark an expected action as undefined in specified circumstances, rather than add an override point that's going to result in copy/pasting code around when other jobs want to do this.

Reply:
Trying to do something like this I think would end up meaning deeper changes to the TrackAction interface and might brush against all the other jobs that use it. At this point I'd rather leave the one-off as a one-off and deal with the interface implications if and when someone else wants to do this, or at least as a separate PR from the rest of this.